AK/AsyncStream.h

```diff
@@ -0,0 +1,288 @@
+        if (!m_is_reading_peek) {
+            m_is_reading_peek = true;
+            auto data = buffered_data_unchecked({});
+            if (!data.is_empty())
+                co_return PeekOrEofResult { data, false };
+        }
+        bool is_not_eof = CO_TRY(co_await enqueue_some({}));
+        co_return PeekOrEofResult { buffered_data_unchecked({}), !is_not_eof };
+    }
+
+    Coroutine<ErrorOr<ReadonlyBytes>> peek()
+    {
+        auto [data, is_eof] = CO_TRY(co_await peek_or_eof());
+        if (is_eof) {
+            reset();
+            co_return Error::from_errno(EIO);
+        }
+        co_return data;
+    }
+
+    Coroutine<ErrorOr<ReadonlyBytes>> read(size_t bytes)
+    {
+        m_is_reading_peek = false;
+
+        if (bytes) {
+            auto buffer = buffered_data();
+            while (buffer.size() < bytes) {
+                if (!CO_TRY(co_await enqueue_some({}))) {
+                    reset();
+                    co_return Error::from_errno(EIO);
+                }
+                buffer = buffered_data_unchecked({});
+            }
+            dequeue({}, bytes);
+            co_return buffer.slice(0, bytes);
+        } else {
+            co_return Bytes {};
+        }
+    }
+
+    virtual Coroutine<ErrorOr<bool>> enqueue_some(Badge<AsyncInputStream>) = 0;
+    virtual ReadonlyBytes buffered_data_unchecked(Badge<AsyncInputStream>) const = 0;
+    virtual void dequeue(Badge<AsyncInputStream>, size_t bytes) = 0;
+
+    template<typename T>
+    Coroutine<ErrorOr<T>> read_object()
+    {
+        auto bytes = CO_TRY(co_await read(sizeof(T)));
+        union {
+            T object;
+            char representation[sizeof(T)];
+        } reinterpreter = {};
+        memcpy(&reinterpreter, bytes.data(), sizeof(T));
+        co_return reinterpreter.object;
+    }
+
+protected:
+    static Badge<AsyncInputStream> badge() { return {}; }
+
+    bool m_is_reading_peek { false };
+};
+
+class AsyncOutputStream : public virtual AsyncResource {
+public:
+    AsyncOutputStream() = default;
+
+    virtual Coroutine<ErrorOr<size_t>> write_some(ReadonlyBytes buffer) = 0;
+
+    virtual Coroutine<ErrorOr<void>> write(ReadonlySpan<ReadonlyBytes> buffers)
+    {
+        for (auto buffer : buffers) {
+            while (!buffer.is_empty()) {
+                auto nwritten = CO_TRY(co_await write_some(buffer));
+                buffer = buffer.slice(nwritten);
+            }
+        }
+        co_return {};
+    }
+};
+
+class AsyncStream
+    : public AsyncInputStream
+    , public AsyncOutputStream {
+public:
+    AsyncStream() = default;
+};
+
+template<typename T>
+class StreamWrapper : public virtual AsyncResource {
+public:
+    StreamWrapper(NonnullOwnPtr<T>&& stream)
+        : m_stream(move(stream))
+    {
+    }
+
+    void reset() override { return m_stream->reset(); }
+    Coroutine<ErrorOr<void>> close() override { return m_stream->close(); }
+    bool is_open() const override { return m_stream->is_open(); }
```

Comment:
What is the purpose of wrapping here? The usages seem as if `StreamWrapper` might as well be omitted.

Reply:
Conceptually, if a class (like `HTTP::Http11Connection`) has an owning reference of a stream as a member while not being a stream by itself, the said class needs to be an `AsyncResource` with these methods implemented to do async "RAII". So, `StreamWrapper` is just a helper to implement those on classes which have one async stream as a member.